staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types_discovery.go

```diff
@@ -0,0 +1,121 @@
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// APIGroupDiscoveryList is a resource containing a list of APIGroupDiscoveries.
+// The shape is chosen to be very similar to "normal" api endpoints and work well with our type system
+// This is what is returned from the /discovery/v2 endpoint.
+type APIGroupDiscoveryList struct {
+	TypeMeta `json:",inline"`
+	// ResourceVersion will not be set, because this does not have a replayable ordering among multiple apiservers.
+	// More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#metadata
+	// +optional
+	ListMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+	// Groups is the list of groups for discovery.
+	// +listType=map
+	// +listMapKey=name
+	Groups []APIGroupDiscovery `json:"groups" protobuf:"bytes,2,rep,name=groups"`
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// APIGroupDiscovery holds information about which resources are being served for all version of the API Group.
+type APIGroupDiscovery struct {
+	TypeMeta `json:",inline"`
+	// Standard object's metadata.
+	// The only field completed will be Name.  For instance, resourceVersion will be empty.
+	// Name is the name of the API group whose discovery information is presented here.
+	// Name is allowed to be "" to represent the legacy, ungroupified resources.
+	// This is included to make the standard deserialization easier.
+	// More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#metadata
+	// +optional
+	ObjectMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+	// Versions are the versions supported in this group. They are sorted in descending order of preference,
+	// with the preferred version being the first entry.
+	// +listType=map
+	// +listMapKey=version
+	Versions []APIVersionDiscovery `json:"versions,omitempty" protobuf:"bytes,2,rep,name=versions"`
+}
+
+// APIVersionDiscovery holds a list of APIResourceDiscovery types that are served for a particular version within an API Group.
+type APIVersionDiscovery struct {
+	// Version is the name of the version within a group version.
+	Version string `json:"version" protobuf:"bytes,1,opt,name=version"`
+	// Resources is a list of APIResourceDiscovery objects for the corresponding group version.
+	// +listType=map
+	// +listMapKey=resource
+	Resources []APIResourceDiscovery `json:"resources,omitempty" protobuf:"bytes,2,rep,name=resources"`
+	// Stale marks whether a group version's discovery document is up to date.
```

Comment:
What values can a client expect?  Human readable message explaining why?

Reply:
empty (false) or true depending on whether the discovery document was reachable. I'm less inclined to put in a longer human readable message simply because it will be mostly machine clients consuming these and also the fact that they'll be cached. Would you be alright if we used this field as a boolean and humans can reference the documentation to understand the values?

Reply:
I swear that was a string a moment ago.  We avoid bools in the API due to how often they aren't actually bools.  I'm ok with a string with enumerate values.  Something like a Status string with values "Current" and "Stale" to start.

Reply:
Updated to `DocumentStatus` with the enum values being `Current` and `Stale`